### PR TITLE
Feature/bags saver

### PR DESCRIPTION
**Pull request by r0mainK**
- Solved [issue 33](https://github.com/src-d/apollo/issues/33) by moving back logic
- Refactored the get_parser a bit: moved the subparsers to match expected order of utilization and seperated them nicely, made `engine` version dynamic  like in `ml`, took out `add_cassandra_args` from `add_templates_args` (confusing when debugging and did not bring much)
- Solved two non reported bug I was having (both `SyntaxError`):
    - In `stream_template` function of `query.py` file, simply switched both arguments
    - In `__call__` method of `CommunityDetector`  in `graph.py` file, added config to `kwarg` dict with `update` methode and removed it from the call to one of the native methods to detect communities of igraph's `Graph` class

EDIT: `dumpcmd` was not working anymore, the query done on the meta table was returning nothing so the communities part of the template was empty. Found that it was bc the `WHERE`clause was querying on the sha1 column, but the values we inputted in the first place were of the form 'reponame@sha1'. Solved it by splitting the strings in the approriate generator 

Error logs, respectively for `query.py`:
```
Traceback (most recent call last):
  File "/usr/local/bin/apollo", line 11, in <module>
    load_entry_point('apollo==0.1.0', 'console_scripts', 'apollo')()
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 480, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 2693, in load_entry_point
    return ep.load()
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 2324, in load
    return self.resolve()
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 2330, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/lib/python3.4/dist-packages/apollo/__main__.py", line 16, in <module>
    from apollo.graph import find_connected_components, dumpcc, detect_communities, dumpcmd, \
  File "/usr/local/lib/python3.4/dist-packages/apollo/graph.py", line 18, in <module>
    from apollo.query import weighted_jaccard, stream_template
  File "/usr/local/lib/python3.4/dist-packages/apollo/query.py", line 105
    template.stream(**kwargs, format_url=format_url).dump(dest)
                            ^
SyntaxError: invalid syntax
```

And for `graph.py`:
```
Traceback (most recent call last):
  File "/usr/local/bin/apollo", line 11, in <module>
    load_entry_point('apollo==0.1.0', 'console_scripts', 'apollo')()
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 480, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 2693, in load_entry_point
    return ep.load()
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 2324, in load
    return self.resolve()
  File "/usr/local/lib/python3.4/dist-packages/pkg_resources/__init__.py", line 2330, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/lib/python3.4/dist-packages/apollo/__main__.py", line 16, in <module>
    from apollo.graph import find_connected_components, dumpcc, detect_communities, dumpcmd, \
  File "/usr/local/lib/python3.4/dist-packages/apollo/graph.py", line 282
    result = action(**kwargs, self.config)
                            ^
SyntaxError: invalid syntax
```

### CODE REVIEW

File: apollo/__main__.py [hunk 1]
```diff
@@ -2,11 +2,13 @@
 import json
 import logging
 import sys
+import pip
 from time import time
 
 from igraph import Graph
 from modelforge.logs import setup_logging
 from sourced.ml import extractors
+from sourced.ml.__main__ import ArgumentDefaultsHelpFormatterNoNone
```
Comment (vmarkovtsev) [inline apollo/__main__.py]: We need to move `ArgumentDefaultsHelpFormatterNoNone` to `sourced/ml/cmd_entries/args.py` and put it into `sourced/ml/cmd_entries/__init__.py`

File: apollo/__main__.py [hunk 2]
```diff
@@ -206,14 +221,12 @@ def add_dzhigurda_arg(my_parser):
     add_spark_args(evalcc_parser, default_packages=[CASSANDRA_PACKAGE])
     add_cassandra_args(evalcc_parser)
 
-    # TODO: retable [.....] -> [.] [.] [.] [.] [.]
```
Comment (vmarkovtsev) [inline apollo/__main__.py]: This TODO still holds: it is about taking the extracted features and building the new hashtables with different parameters. Existing code does it in `hash` command. We probably just need the command to load the features from Cassandra and store them as batch ASDF files in the same format as `bags`.

File: apollo/bags.py [hunk 3]
```diff
@@ -86,9 +88,60 @@ def preprocess_source(args):
...
 def source2bags(args):
+    log = logging.getLogger("sourced2bags")
+    extractors = create_extractors_from_args(args)
+    session_name = "sourced2bags-%s" % uuid4()
     cassandra_utils.configure(args)
-    return repos2bow_entry_template(
-        args,
-        select=lambda: DzhigurdaFiles(args.dzhigurda),
-        cache_hook=lambda: MetadataSaver(args.keyspace, args.tables["meta"]))
+    if args.parquet:
```
Comment (vmarkovtsev) [inline apollo/bags.py]: We cannot explode this. The code must re-use the template. Think how we can extend it with the **minimal** changes.

File: apollo/graph.py [hunk 4]
```diff
@@ -279,7 +279,8 @@ def __call__(self, graph):
             kwargs = {"weights": graph.edge_weights}
         if self.algorithm == "edge_betweenness":
             kwargs["directed"] = False
-        result = action(**kwargs, **self.config)
+        kwargs.update(self.config)
+        result = action(**kwargs)
```
Comment (vmarkovtsev) [inline apollo/graph.py]: Please put TODO here which explains that Python 3.4 sucks. Once we move to Ubuntu 18.04, we stop supporting 14.04 and thus 3.4 and remove this hack.

**Comment by r0mainK** [comment]
Ok, will separate this PR into 4 different PRs

File: apollo/__main__.py (hunk 2 above)
Comment (r0mainK) [reply apollo/__main__.py]: My bad, will put it back

File: apollo/bags.py (hunk 3 above)
Comment (r0mainK) [reply apollo/bags.py]: Okay. Then I could simply add a `save_bags_hook` to the template, like is done to save the metadata. Will create a PR tomorrow in ml.

**Pull request closed without merging**